Novo Layout: Morpheus

megatroom opened the pull request:
<img align="left" width="120" height="105" src="https://user-images.githubusercontent.com/673904/60402949-0a6a7400-9b6d-11e9-82ef-8f0320ad01ab.jpg">
"This is your last chance. After this, there is no turning back. You take the blue pill - the story ends, you wake up in your bed and believe whatever you want to believe. You take the red pill - you stay in Wonderland and I show you how deep the rabbit-hole goes."

# The red pill

Now that you've entered the Matrix, here's the new layout:

![office](https://user-images.githubusercontent.com/673904/60403025-f6734200-9b6d-11e9-8149-a201bdd78ead.png)

I used the concepts of Material Design and is fully responsive.

You can see all online users through the side menu:

![office-sidebar](https://user-images.githubusercontent.com/673904/60403060-9c26b100-9b6e-11e9-98d2-c53f60e2ad41.png)

The side menu has a search for you can find a user more quick.

Now the video conference is part of the page, so it is possible to take advantage of the menu for the room's own functions:

![call](https://user-images.githubusercontent.com/673904/60403083-10615480-9b6f-11e9-821b-4e8b4094b77f.png)

And you still open the side menu:

![call-sidebar](https://user-images.githubusercontent.com/673904/60403099-6635fc80-9b6f-11e9-8e1f-cd7046628181.png)

You can use the side menu to invite some user to your current meeting:

![invite](https://user-images.githubusercontent.com/673904/60403111-8cf43300-9b6f-11e9-83da-b958d4cf995d.png)

The user will receive the invitation to join the meeting.

If you want to find a new meeting room fast, you can click in a menu filter to show only the rooms with people:

![active-filter](https://user-images.githubusercontent.com/673904/60403160-2885a380-9b70-11e9-8853-8c92075904f0.png)

You can search by the name instead:

![search](https://user-images.githubusercontent.com/673904/60403164-3fc49100-9b70-11e9-8c3d-e15c7a584709.png)

If new users join a meeting, you will recieve a notification:

![notifications](https://user-images.githubusercontent.com/673904/60403170-579c1500-9b70-11e9-8281-45af5aa5cb05.png)

# Backward Compatibility

This new layout is only a proposal, although it already has a lot ready, this new version does not affect the current version, can be used together.

By default you will enter the room that exists today, but if you follow the white rabbit you can use this new version.

> Follow the White Rabbit...

There are a few things to adjust, both in code and functionality, but if the proposal is interesting to you, I give continuity to the development.

I hope you all liked :grin:

alesshh commented:
Awesome @megatroom! 
I will review and test asap.
"There is no way back from here"
https://www.youtube.com/watch?v=fTaOlBWcl48

megatroom commented:
> Hi @megatroom you are rock \oo/. I'm executing basic tests and I will put my comments.
> 
> 1. one of the main features is to show where you are available. I noticed that in this suggestion every time when I left the meeting my user is redirected to the default room. The rooms is a good way to identify team and areas. And sometimes I only want to show that I'm online working with my team. You can see this use case in this print:
>    ![image](https://user-images.githubusercontent.com/643779/60473937-d0b96c00-9c45-11e9-99bc-ed758e252fcd.png)

I thought it was a bug staying in the room after leaving the Jitsi, I had understood that the fact of being in the room the user would be mandatory in the meeting (Jitsi).

Now it makes more sense, and it makes sense now to have the option to get into meeting in the standard room, and now it made sense of this headset icon.

This was not clear because there is only one button to enter the room, and when you enter it automatically enters the Jitsi. It would make more sense to have two buttons then: "Enter room" and "Enter meeting". What do you think?

juliemar commented:
Hi @megatroom sometimes I need to send the link of my current room to other users, and I miss the room link in this propose:

**Current version:**
![image](https://user-images.githubusercontent.com/643779/60531382-41f23100-9cd1-11e9-8230-84be2e7360a0.png)

**Morpheus Version:**
![image](https://user-images.githubusercontent.com/643779/60531484-7cf46480-9cd1-11e9-8edb-cfdf0e338ee7.png)

I'm in room logos. In this case, is not possible to send the room link for other users to enter directly in my current room.

juliemar commented:
> This was not clear because there is only one button to enter the room, and when you enter it automatically enters the Jitsi. It would make more sense to have two buttons then: "Enter room" and "Enter meeting". What do you think?

I agree with you, the current UX is not good. by the way, is so important to pin my user in room of my team for example. This feature enables other people to quickly identify the team and who is working with this treams.

megatroom commented:
> Hi @megatroom sometimes I need to send the link of my current room to other users, and I miss the room link in this propose:
> I'm in room logos. In this case, is not possible to send the room link for other users to enter directly in my current room.

It makes perfect sense. I'll implement that.

> I agree with you, the current UX is not good. by the way, is so important to pin my user in room of my team for example. This feature enables other people to quickly identify the team and who is working with this treams.

What do you mean by pin the user? Do you wanna group the users in the sidebar by room, or do you want to mark the user as a favorite or put a tag so it goes to the top of the list?

juliemar commented:
> What do you mean by pin the user? Do you wanna group the users in the sidebar by room, or do you want to mark the user as a favorite or put a tag so it goes to the top of the list?

I'm sorry about my mistake. Is important to enter in the room and stay there. Is not necessary to be in the call to enter the room. the room concept is to inform where you are available. because this is important to be in a room even without being on a call.

megatroom commented:
@juliemar tasks done and committed:

- [x] I've changed the route to include room's ID so it can be shared.
- [x] I added a button in app bar to share room's link more easily.
- [x] I divided the card's buttons to "enter room" and "enter meeting" to have that difference between changing rooms and entering the call.
- [x] If the user clicks the end button on the Jisti, it is redirected to main page.

Pending:

- [ ] Visually mark users who are in a meeting.
- [ ] Fix users side menu to not show invite button for users already in current meeting.
- [ ] Include some unit tests

Let me know if there's anything else.

juliemar commented:
The last one ;) 
When I only enter in room the URL is not changing with the unique roomId

After this, for me merge and Go :)

megatroom commented:
> The last one ;)
> When I only enter in room the URL is not changing with the unique roomId
> 
> After this, for me merge and Go :)

I can't reproduce this bug, for me whenever I enter the room the room id is in the URL.

![Screenshot from 2019-07-05 12-16-28](https://user-images.githubusercontent.com/673904/60731787-d56c7180-9f1e-11e9-90a1-32f3048c3026.png)

Can you send me some evidence? What is the list of rooms you use?

juliemar commented:
I made a video: https://drive.google.com/file/d/1F2BLQfJstNQQCpzSS4LejH9QxkYrMNWV/view

my expectation is: when I enter the room the URL changes like when I enter in call.. maybe we can have two routes.
1) room meet link as you did (it was very good)
2) room link like I'm trying to show in my video

What do you think?

megatroom commented:
> I made a video: https://drive.google.com/file/d/1F2BLQfJstNQQCpzSS4LejH9QxkYrMNWV/view
> 
> my expectation is: when I enter the room the URL changes like when I enter in call.. maybe we can have two routes.
> 
> 1. room meet link as you did (it was very good)
> 2. room link like I'm trying to show in my video
> 
> What do you think?

Now I got it, you wanna in the main screen, where it has the room grid, the URL is with the id of the room that you are.
Okay, I'm going to implement this route.

megatroom commented:
@juliemar , I've finished all changes.

Let me know if everything is ok.

juliemar commented:
1) When you are not logged and try to access the morpheus URL the we have a redirect loop.
video: https://drive.google.com/file/d/11oALrio7Xyp_itIucHX4SJeZ_w_0Vj33/view

2) Web notifications: In this propose the web notifications notify all other users room change. We tried this in the first version and had to change because the users don't know the other user's movement. They only want to know if a user enters or left their current room. And I missed the web notification in desktop. is important to be desktop because some times I'm using other software or have an office minimized and have to know if someone is calling or entered in my current room.
video: https://drive.google.com/file/d/1VzmzErr_wNrpp7hdF7RjXtyJlwRUnbju/view

this is the commits that I mentioned in my video.
https://github.com/ResultadosDigitais/matrix/commit/d8dd46c6d7ccb2abdfbd30fb4192375bdb8623ba
https://github.com/ResultadosDigitais/matrix/commit/2345dbd50a45b7fecfcf5e6c38ae085179e8fea7

megatroom commented:
> 1. When you are not logged and try to access the morpheus URL the we have a redirect loop.
>    video: https://drive.google.com/file/d/11oALrio7Xyp_itIucHX4SJeZ_w_0Vj33/view

- I resolved the infinite loop.
- I was relying on the localStorage to set whether the user was logged in, I changed this and placed to validate the login by the Google API. In addition to getting more consistent, I reload the localStorage if it is empty.

> 2. Web notifications: In this propose the web notifications notify all other users room change. We tried this in the first version and had to change because the users don't know the other user's movement. They only want to know if a user enters or left their current room. And I missed the web notification in desktop. is important to be desktop because some times I'm using other software or have an office minimized and have to know if someone is calling or entered in my current room.

- I changed to only display the notification for when a user enters the same room as you are.
- I added a button in the menu to disable the notification in case you do not want to receive them temporarily.
- A browser's notification was already being sent in this case, what may be happening is that you may not have allowed the notification. To make it easier to identify this problem, if the notification is blocked or if the user has not yet made this decision, the notification icon I added will get an exclamation mark, and clicking on it will attempt to enable or be warned that the user has blocked the notification.

juliemar commented:
@alesshh and @lccezinha for me is everything OK. I only reviewed the functionalities, about code I will let for you :)

alesshh commented on frontend/morpheus/store/reducers.js:
```diff
@@ -0,0 +1,224 @@
+        }
+      };
+    case ADD_USER: {
+      const index = state.usersInRoom.findIndex(
```
My suggestion is use a map instead array to store users per room. It's reduce complexity from o(n) to o(1) and simplify add / remove users.

```
const usersInRoom = {}
usersInRoom[action.user.id] = { user: action.user, room: roomId }  } 
Object.values(usersInRoom) // [{user: ...}, {user: ...}]
...
```

alesshh commented on frontend/morpheus/store/selectors.js:
```diff
@@ -0,0 +1,19 @@
+
+export const selectError = state => state.error;
+
+export const selectSettings = state => state.settings;
```
I think this file generate an unnecessary dependency and complexity.
My suggestion is:
- use `mapStateToProps` only where necessary
- rely in default props with explicit required

megatroom replied:
I don't understand, could you explain?

megatroom commented:
> Hey @megatroom , sorry for the delay in review.
Hello @alesshh ! Thanks for the review.

> I added some comments, if imakes sense.
> I has 2 concerns:
> 
> * I think we can be restrict with props, using optional / default props only where is necessary, leaving props less verbose

When it's too restrictive to output contracts in front-end it ends up reducing the application resilience. The default values are a guarantee for the component to be rendered even in the absence of any value. It's a part of defensive programming.

To improve props validation, I came back with ESLint's _react/forbid-prop-types_ rule and I specified each array and object better.

What do you think?

> * Use of mocha and jest: I don't have personal preferences, but it would be nice if we can live only with one test framework, reducing complexicity / dependencies. We can kill jest depencendy after?

I removed the Jest and left only Mocha testing everything.

I really like Mocha, it's even my first choice for back-end testing. However, I like Jest to test front-end, mainly projects with React, since it already comes ready with jsdom, and come with assertion library and mock by default.

Apart from personal preference for Jest, I also find it a little strange to have the back-end and front-end tests all together in the same directory. Besides it's looks like a unnecessary coupling, it's difficult to configure, as for example I had to repeat the ESLint configuration.

> Other think is about the approach of structure by function instead feature (https://medium.com/better-programming/scaling-your-redux-app-with-ducks-6115955638be). I like the approach by feature, because I can easy locate and isolate things by context, and easy identify shared components. Mainly in medium/large apps this structure tends to get confused, but for now, we can live with this.

I like modularization by functionality, and also Redux Ducks, this was for a long time my preferred form of organization, but currently I have followed [this pattern of organization for any project](https://react-file-structure.surge.sh).

If you look at the history of commits I made, you will see that I started using React only with useReducer in App.js to control the global state, with the files and component all in the same directory as I was evolving, I was organizing and adding more things, the Redux I put practically at the end, when I had already opened the PR.

We can organize the files on demand in the next PR's.

megatroom commented on frontend/components/MenuUsers.js:
```diff
@@ -0,0 +1,145 @@
+          !(user.inMeet && user.roomId === currentRoom.id);
+
+        return (
+          <ListItem key={user.id}>
```
I agree, it turned out that this component was quite large. I did as you suggested.

megatroom commented on frontend/components/NotificationCheckbox.js:
```diff
@@ -0,0 +1,74 @@
+      </Tooltip>
+    );
+  }
+
```
`NotificationCheckbox` is already a subset of `MenuOffice` component to concentrate the logic of menu notification button. And the `"@material-ui"` already encapsulates the presentation layer of the components, you may notice that there is no style here, just behavior. So my fear is that with this division it makes maintenance difficult without having any gain in return. What do you think?

megatroom commented on frontend/morpheus.js:
```diff
@@ -0,0 +1,51 @@
+  gapi.load("auth2", () => {
+    gapi.auth2.init().then(
+      auth2 => {
+        if (auth2.isSignedIn.get()) {
```
I switched to early return, but I can't put the return on the same line because it's an assignment, the lint complains about it.

megatroom commented on frontend/morpheus/App.js:
```diff
@@ -0,0 +1,360 @@
+    const profile = initProfile();
+
+    if (profile.isProfileStored()) {
+      onSetCurrentUser(getCurrentUser());
```
The `useSocket` and `useEvents` are custom [React Hooks](https://reactjs.org/docs/hooks-intro.html) and not HOC. But it makes perfect sense to separate them into files and I did it.

The organization of this hooks it's a little confusing, I agree. I started developing the logic inside the useEffect, it got bigger, so I separated in a custom hook, and after it was bigger I saw that it was able to separate into two different custom hooks. But there's still a lot of behavior coupled with them.

I've tried to separate the maximum of chat logic in `socket.js`, but as **Morpheus is a proposal**, I tried to do as little as possible in what already existed, in case you did not like it would be easy after removing it. So I ended up leaving some technical debts.

Since you all like Morpheus, we can see a better way to organize it when it's replacing the old front end.

megatroom commented on frontend/morpheus/App.js:
```diff
@@ -0,0 +1,360 @@
+        }
+      }, 500);
+
+      events.onSyncOffice(usersInRoom => {
```
Done!

megatroom commented on frontend/morpheus/containers/OfficeRedirect.js:
```diff
@@ -0,0 +1,42 @@
+  useEffect(() => {
+    if (currentRoom && currentRoom.id) {
+      setRoomId(currentRoom.id);
+    } else if (rooms.length > 0) {
```
I can't use early return inside the hook useEffect because the return of your callback is used for its cleanup.

megatroom commented on frontend/morpheus/containers/RoomPage.js:
```diff
@@ -0,0 +1,157 @@
+      } else {
+        setLoading(false);
+        setMeetingDialogOpen(true);
+      }
```
I can't use early return inside the hook useEffect because the return of your callback is used for its cleanup.
